### ISSUE TITLE
perf: Investigate splitting TimestampCache into read-only and read-write interval caches

### ISSUE THREAD

**Issue by nvanbenschoten** [cockroachdb/cockroach#4700]
Related to #4616 and should wait on #4699.

Currently we always make the distinction between read and write timestamp interval values when inserting and querying the `IntervalCache` within the `TimestampCache`. Because of this, we should investigate splitting the single `IntervalCache` into a pair of **read-only** and a **read-write** `IntervalCache`s. Splitting up the underlying interval tree into two smaller trees should allow for improved performance during insertion, deletion, and most importantly querying (`GetOverlaps()`).

**Comment by tschottdorf** [comment]
The write timestamp is a little weird anyway. A cursory inspection suggests that it's never really used; the usage in `(*Replica)` is the follows:

```go
				// If there's a newer write timestamp...
				if !wTS.Less(ba.Timestamp) {
					// If we're in a txn, we still go ahead and try the write since
					// we want to avoid restarting the transaction in the event that
					// there isn't an intent or the intent can be pushed by us.
					//
					// If we're not in a txn, it's trivial to just advance our timestamp.
					if ba.Txn == nil {
						ba.Timestamp = wTS.Next()
					}
				}

```

so the write timestamp is essentially ignored unless we're not in a transaction (which is rare). So one upshot of splitting them is that the write tree would basically never get queried.

As for the splitting itself, is it obviously more efficient? Isn't querying `O(#overlaps + log(#intervals))`? That means splitting doesn't change asymptotics for the overlap portion, and the `log(#intervals)` is better than `log(#readintervals) + log(#writeintervals)`.

**Comment by nvanbenschoten** [comment]
Based off the code above, we should probably split `GetMax` into `GetMaxRead` and `GetMaxWrite` so that we don't even need to look at writes unless we're not in a transaction.

In terms of the querying efficiency, I'm worried much more about the query that shows up during profiling because of how hot a code path it is in, that being the query for overlapping intervals in `(*TimestampCache).Add`. Right now, we query the entire tree for overlaps, and always throw out all either all of the reads or all of the writes found immediately, which is a complete waste. This also makes the insertion later on in that method slower, and deletions in the new PR.

```go
for _, o := range tc.cache.GetOverlaps(key.Start, key.End) {
    ce := o.Value.(*cacheValue)
    if ce.readOnly != readOnly {
        continue
    }
    ...
}
```

**Comment by tschottdorf** [comment]
Ah, that makes sense. Looks like the splitting is the way to go, especially since we won't query the write tree at all.
